File: byzantine_open_world_power_v1_public/byzantine_open_world_power_v1/verify.py
from __future__ import annotations

from collections import Counter
from pathlib import Path
import hashlib
import json

from .engine import digest

ROOT = Path(__file__).resolve().parents[1]
REPORTS = ROOT / "reports"


class VerificationError(ValueError):
    pass


def _read(path: Path):
    return json.loads(path.read_text(encoding="utf-8"))


def _file_sha(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def _require(condition: bool, message: str) -> None:
    if not condition:
        raise VerificationError(message)


def verify(root: Path = ROOT) -> dict:
    reports = root / "reports"
    contract = _read(root / "contract.json")
    rows = _read(reports / "rows.json")
    summary = _read(reports / "summary.json")
    receipt = _read(reports / "receipt.json")

    _require(receipt.get("schema") == "byzantine-open-world-power-v1/public-receipt/1", "receipt schema")
    _require(digest(receipt["payload"]) == receipt["sha256"], "receipt self-hash")
    _require(len(rows) == contract["scenario_count"] == 72, "row count")
    _require(summary["status"] == "PASS", "summary status")
    _require(summary["scenario_count"] == 72 and summary["pass_count"] == 72, "summary counts")
    _require(all(row.get("pass") is True for row in rows), "row pass")
    _require(digest(rows) == receipt["payload"]["rows_sha256"], "rows hash")
    _require(digest(summary) == receipt["payload"]["summary_sha256"], "summary hash")
    _require(_file_sha(root / "contract.json") == receipt["payload"]["contract_sha256"], "contract hash")

    for relative, expected in receipt["payload"]["source_sha256"].items():
        _require(_file_sha(root / relative) == expected, f"source hash: {relative}")

    expected_ids = {f"{domain}::{archetype}" for domain in contract["domains"] for archetype in contract["archetypes"]}
    actual_ids = {row["scenario_id"] for row in rows}
    _require(len(actual_ids) == len(rows), "duplicate scenario")
    _require(actual_ids == expected_ids, "scenario matrix completeness")

    counts = dict(sorted(Counter(row["terminal"] for row in rows).items()))
    _require(counts == contract["expected_terminal_counts"], "row terminal counts")
    _require(summary["terminal_counts"] == counts, "summary terminal counts")
    _require(receipt["payload"]["expected_terminal_counts"] == counts, "receipt terminal counts")

    for row in rows:
        expected_terminal = contract["expected_terminal"][row["archetype"]]
        _require(row["terminal"] == expected_terminal, f"terminal: {row['scenario_id']}")
        _require(row["decision"]["terminal"] == row["terminal"], f"decision terminal: {row['scenario_id']}")
        decision_payload = {key: value for key, value in row["decision"].items() if key != "sha256"}
        _require(digest(decision_payload) == row["decision"]["sha256"], f"decision hash: {row['scenario_id']}")

    result = {"status": "PASS", "receipt_sha256": receipt["sha256"], "scenarios": len(rows)}
    print(json.dumps(result, sort_keys=True))
    return result


if __name__ == "__main__":
    try:
        verify()
    except (KeyError, TypeError, VerificationError, json.JSONDecodeError, OSError) as exc:
        raise SystemExit(f"REJECTED: {exc}") from exc

File: byzantine_open_world_power_v1_public/byzantine_open_world_power_v1/verify.js
#!/usr/bin/env node
const fs=require('fs'),path=require('path'),crypto=require('crypto');
const R=path.resolve(__dirname,'..'), P=path.join(R,'reports');
const read=p=>JSON.parse(fs.readFileSync(p,'utf8'));
const canon=v=>Array.isArray(v)?`[${v.map(canon).join(',')}]`:v&&typeof v==='object'?`{${Object.keys(v).sort().map(k=>`${JSON.stringify(k)}:${canon(v[k])}`).join(',')}}`:JSON.stringify(v);
const dig=v=>crypto.createHash('sha256').update(canon(v)).digest('hex');
const fsha=p=>crypto.createHash('sha256').update(fs.readFileSync(p)).digest('hex');
function req(x,m){if(!x)throw Error(m)}
try{
 const c=read(path.join(R,'contract.json')), rows=read(path.join(P,'rows.json')), s=read(path.join(P,'summary.json')), rec=read(path.join(P,'receipt.json'));
 req(rec.schema==='byzantine-open-world-power-v1/public-receipt/1','receipt schema');
 req(dig(rec.payload)===rec.sha256,'receipt self-hash'); req(rows.length===c.scenario_count&&rows.length===72,'row count'); req(rows.every(r=>r.pass===true),'row pass');
 req(s.status==='PASS'&&s.scenario_count===72&&s.pass_count===72,'summary'); req(dig(rows)===rec.payload.rows_sha256,'rows hash'); req(dig(s)===rec.payload.summary_sha256,'summary hash');
 req(fsha(path.join(R,'contract.json'))===rec.payload.contract_sha256,'contract hash');
 for(const [n,h] of Object.entries(rec.payload.source_sha256)) req(fsha(path.join(R,n))===h,`source ${n}`);
 const expectedIds=new Set(c.domains.flatMap(d=>c.archetypes.map(a=>`${d}::${a}`))), ids=new Set();
 const counts={};
 for(const r of rows){
   req(!ids.has(r.scenario_id),'duplicate'); ids.add(r.scenario_id); counts[r.terminal]=(counts[r.terminal]||0)+1;
   req(r.terminal===c.expected_terminal[r.archetype],`terminal ${r.scenario_id}`); req(r.decision.terminal===r.terminal,`decision terminal ${r.scenario_id}`);
   const d={...r.decision}; delete d.sha256; req(dig(d)===r.decision.sha256,`decision hash ${r.scenario_id}`);
 }
 req(ids.size===expectedIds.size&&[...expectedIds].every(id=>ids.has(id)),'scenario matrix completeness');
 const ordered=Object.fromEntries(Object.entries(counts).sort(([a],[b])=>a.localeCompare(b)));
 req(canon(ordered)===canon(c.expected_terminal_counts),'row terminal counts'); req(canon(s.terminal_counts)===canon(ordered),'summary terminal counts'); req(canon(rec.payload.expected_terminal_counts)===canon(ordered),'receipt terminal counts');
 console.log(JSON.stringify({status:'PASS',receipt_sha256:rec.sha256,scenarios:rows.length}));
}catch(e){console.error(`REJECTED: ${e.message}`);process.exit(2)}
